### PR TITLE
HDDS-7275. Improve error logging when blocks cannot be allocated with RATIS THREE.

### PR DESCRIPTION
**Pull request by SaketaChalamchala**
## What changes were proposed in this pull request?

Current implementation returns a null and later on ScmBlockLocationProtocolServerSideTranslatorPB#allocateScmBlock() throws an exception with #blocks allocated and requested.
```
# hdfs dfs -put /etc/krb5.conf ofs://ozone1/vol1/bucket/3
put: Allocated 0 blocks. Requested 1 blocks
```
Proposed changed returns an exception when a block cannot be allocated after trying all possible strategies

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7275

## How was this patch tested?

Manual.
```
sh-4.2$ ozone sh key put /vol1/bucket1/key1 CONTRIBUTING.md
INTERNAL_ERROR Unable to allocate a container to the block of size: 1048576, replicationConfig: RATIS/THREE. Unable to find enough nodes that meet the space requirement of 10485760 bytes for metadata and 1073741824 bytes for data in healthy node set. Required 3. Found 2.
sh-4.2$ 
```

### CODE REVIEW

**Comment by errose28** [comment]
Thanks for improving this @SaketaChalamchala, this message has been confusing for users before. Could we also add a message summarizing why the allocation failed? For example, the log lines farther up in `getContainer` provide more information to the SCM log, but those messages could be returned to the client as well.

**Comment by SaketaChalamchala** [comment]
> Thanks for improving this @SaketaChalamchala, this message has been confusing for users before. Could we also add a message summarizing why the allocation failed? For example, the log lines farther up in `getContainer` provide more information to the SCM log, but those messages could be returned to the client as well.

Done.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java
Comment (errose28) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java:177]: Suggested change:
```
    // we have tried all strategies we know but somehow we are not able
    // to get a container for this block. Log that info and throw an exception.
```

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java
Comment (errose28) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java:141]: Should we include the allocated pipeline IDs in this message too?

**Review by errose28 — COMMENTED** [review]
Functionally LGTM just minor comments remaining.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java
Comment (SaketaChalamchala) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java:141]: Would it provide more clarity than the error message? It might not be efficient to send allocated pipeline list to the client is it is too large.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java
Comment (SaketaChalamchala) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java]: Done.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java
Comment (errose28) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableRatisContainerProvider.java:141]: Sure we can omit it.

**Review by errose28 — APPROVED** [review]
Thanks for fixing this @SaketaChalamchala. CI will likely fail until #5217 is merged (should be soon). I will hold off running CI on this PR until that is resolved.

**Comment by errose28** [comment]
CI failures are fixed by #5217

**Pull request merged**
